Ray Tracing Extended

gpx1000 opened the pull request:
## Description

Render a simple scene showing the basics of ray tracing, including Ambient Occlusion, top and bottom level acceleration structure rebuilding for animation, multiple objects in a scene, and refraction. The sample loads the sponza gltf model, loads the base fire particle texture and creates geometry for one particle and one quad showing heat above fire then creates a bottom acceleration structure for each geometry type, then makes instances of those static geometries and multiple instances of the fire particle to achieve a particle effect.  It then demonstrates TLAS (Top Level Acceleration Structure) animation via updating the fire particle(s) and BLAS (Bottom Level Acceleration Structure) animation via updating the quad geometry.

## Tested Environments

- Windows 10 / Ubuntu Linux
- GeForce 3060 / GeForce 2080 / GeForce 1070
- Vulkan version: 1.2.176

## Sample Assets changes
depends upon flame texture found in https://github.com/KhronosGroup/Vulkan-Samples-Assets/pull/12

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [ ] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

wflohry committed:
Using basic raytracing as template

wflohry committed:
Load sponza

wflohry committed:
Loading sponza model

wflohry committed:
Separating out ray tracing command buffers

wflohry committed:
Add render modes

wflohry committed:
Edit vertex structure

wflohry committed:
Put static objects in buffer with device-exclusive memory

wflohry committed:
Adding vertex and index data to bindings

wflohry committed:
Demonstration of referencing model data from ray tracing shader

wflohry committed:
Implemented texture array and visualization

wflohry committed:
Added fire pit

wflohry committed:
Added ray-traced shadows

wflohry committed:
Adding support for AO

wflohry committed:
Updated model settings

gpx1000 committed:
Fixed a few warnings and spelling errors.

Also fixed the fences problem that causes crash bug from fences on close.

wflohry committed:
Separating out static model loading

wflohry committed:
Separate handling of static / dynamic buffers

wflohry committed:
Added dynamic object and refactored

gpx1000 committed:
Fixed another warning and fixed the fences.

gpx1000 committed:
Merge branch 'feature/raytrace-extended' into warn_removal_and_fence_fix

# Conflicts:
#	samples/extensions/raytracing_extended/raytracing_extended.cpp

gpx1000 committed:
Switch to using constants and specializations in order to change values.  Can switch back to uniforms at a later date but the uniforms were proving a render issue for linux.  This should also demonstrate how to use multiple specialization constants in a ray traced environment.

gpx1000 committed:
more warnings removal

wflohry committed:
Merge warn_remove_and_fence_fix

gpx1000 committed:
Merged in warn_removal_and_fence_fix (pull request #1)

Fixed a few warnings and spelling errors.

Approved-by: William Lohry

wflohry committed:
Merge branch 'feature/raytrace-extended' of ssh://bitbucket.org/holochip/vulkan-samples into feature/raytrace-extended

wflohry committed:
Moving instance buffer creation to TLAS construction

wflohry committed:
Adding refraction placeholder

wflohry committed:
Perform BLAS update for dynamic objects and TLAS update every frame.

wflohry committed:
Adding dynamic sinusoidal wave

wflohry committed:
Finished refractive wave

wflohry committed:
Update to refraction

wflohry committed:
Adding separate flame instance

wflohry committed:
Working flame / particle generator

wflohry committed:
Updates to flame and refraction calculation

wflohry committed:
Updated refraction algorithm

wflohry committed:
Added texture

gpx1000 committed:
more warnings removal
move the png to ktx format

gpx1000 committed:
ensure we're pointing to the right place

wflohry committed:
Merge and update

wflohry committed:
Update to ktx commit

gpx1000 committed:
run clang-format over the sample.

wflohry committed:
Added transformations

wflohry committed:
Merged

wflohry committed:
Merged

gpx1000 committed:
Merged in clang-format (pull request #3)

run clang-format over the sample.

Approved-by: William Lohry

wflohry committed:
Adding transformations

wflohry committed:
Implemented automatic  billboard rotation

wflohry committed:
Merge branch 'feature/raytrace-extended' of ssh://bitbucket.org/holochip/vulkan-samples into feature/raytrace-extended

gpx1000 committed:
run a profile pass.
move pts, uvs, normals, indices to be created/initialized once to avoid dynamic memory generation.
remove warnings.
Update copyright.

gpx1000 committed:
Merged in profile (pull request #4)

run a profile pass.

Approved-by: William Lohry

gpx1000 committed:
Update copyright in CMakeLists.txt
Update sample cmake function definition
Update samples list in Samples CMakeLists.txt
Update README.md with first pass definition of the sample.

gpx1000 committed:
Merged in draft_ready (pull request #5)

Update copyright in CMakeLists.txt

Approved-by: William Lohry

wflohry committed:
Added tutorial rough draft

gpx1000 committed:
grammar check.

gpx1000 committed:
grammar check.

gpx1000 committed:
Merged in tutorial (pull request #6)

grammar check.

Approved-by: William Lohry

wflohry committed:
Added fence reset

wflohry committed:
Updating triangle indexing code

wflohry committed:
Avoid dynamic allocation / copy

gpx1000 committed:
update copyright

gpx1000 committed:
Merged in feature/raytrace-extended (pull request #7)

Feature/raytrace extended

gpx1000 committed:
Merge branch 'downstream-master'

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/KhronosGroup/Vulkan-Samples?pullRequest=313) <br/>All committers have signed the CLA.

gpx1000 committed:
Fixes for Clang-format and copyright headers CI failures.

cforfang commented on shaders/khr_ray_tracing_extended/closesthit.rchit:
```diff
@@ -0,0 +1,168 @@
+        return; // this shouldn't happen
+      }
+      // obtain texture coordinate
+      vec4 tex_value = texture(textures[imageOffset], texcoord);
```
Should `imageOffset` be wrapped by `nonuniformEXT` here? Also, are derivatives valid -- or should this be using `textureLod`?

gpx1000 replied:
derivatives I believe are valid, or I at least have no reason to think otherwise.  This isn't using mipmaps for the main reason that I didn't want to make the sample too complicated.  Meaning if I add mipmaps, then why not add don't update things not in view?  or update the particle effect TLAS to use less particles if the LOD is far away etc.

I agree it should be nonuniformEXT.  I'll submit another commit with it corrected, thanks!

cforfang replied:
Cheers :) If there's no mips, I would maybe argue using `textureLod(..., 0)` would be better -- as there's a risk people will copy this code as it _is_ sample code. I somewhat doubt derives are valid TBH, so my guess is this likely only works properly as long as there's only 1 mip.

gpx1000 replied:
That's actually a very good point.  I'll switch to textureLod and leave a comment that spells out reasoning.

Thanks!

gpx1000 committed:
add in nonuniformEXT to the texture lookup as imageOffset isn't uniform.

gary-sweet reviewed (commented):
Builds ok for me, and runs with correct warnings about being unsupported.
I can't actually test the raytracing code though.

gpx1000 committed:
switch to textureLod from texture to ensure it's ready for mipmaps if copied from sample.

SaschaWillems commented:
Great looking sample. Runs fine on my RTX 2060.

One thing I noticed, is that the heat haze effect doesn't look correct when viewed from different angles. E.g. if you watch the fire from the other side, it looks like this:

![image](https://user-images.githubusercontent.com/10283127/128598156-eaf348ec-8e27-4516-997f-e2a1f421bbcd.png)

It would also be great of alpha masking was applied on alpha masked materials:

![image](https://user-images.githubusercontent.com/10283127/128598184-6c512ae0-0419-40e1-9cef-f1c9363aa63c.png)

Afair this would require use of an additional any hit shader.

gpx1000 commented:
Thanks @SaschaWillems 

Good observations on refraction.  It kinda looks like the image is getting flipped on the screen shot.

For the alpha, I thought about the benefits of adding it, but worry about making the sample too complex to serve the purpose of highlighting AO, animation, and shadows.  I think we also need to update the textures the sponza model is using as I don't think it currently has the alpha mask; looking at the ktx files it seems none of them have the alpha baked in either.

wflohry committed:
Fixing refraction bug

gpx1000 committed:
Merge remote-tracking branch 'downstream/feature/raytrace-extended'

SaschaWillems commented:
Refraction now looks fine from all angles.

I'm also more than fine with not having alpha, esp. with the model needing to be updated.

So this looks good to me now :)

SaschaWillems commented on samples/README.md:
````diff
@@ -185,6 +185,14 @@ Uses the debug utilities extension to name and group Vulkan objects (command buf
 
+### [Extended hardware accelerated ray tracing](./extensions/raytracing_extended)<br/>
+**Extensions**: [```VK_KHR_ray_tracing_pipeline```](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_KHR_ray_tracing_pipeline), [```VK_KHR_acceleration_structure```](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_KHR_acceleration_structure)<br/>
+Render Sponza with Ambient Oclusion.  Place a vase in center.  Generate a particle fire that 
````
Small typo. Should read "occlusion".

SaschaWillems commented on samples/README.md:
````diff
@@ -185,6 +185,14 @@ Uses the debug utilities extension to name and group Vulkan objects (command buf
+### [Extended hardware accelerated ray tracing](./extensions/raytracing_extended)<br/>
+**Extensions**: [```VK_KHR_ray_tracing_pipeline```](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_KHR_ray_tracing_pipeline), [```VK_KHR_acceleration_structure```](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VK_KHR_acceleration_structure)<br/>
+Render Sponza with Ambient Oclusion.  Place a vase in center.  Generate a particle fire that 
+demonstrates the TLAS (Top Level Acceleration Structure) animation for the same underlining geometry.
````
Not a native speaker, but I think this should be "underlying" instead of "underlining"?

gpx1000 replied:
You're correct :)  it's updated now.  Good catch.

SaschaWillems commented on samples/extensions/raytracing_extended/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+    CATEGORY ${CATEGORY_NAME}
+    AUTHOR "Holochip Corporation"
+    NAME "Ray tracing extended"
+    DESCRIPTION "Extended example of Ray Tracing highlighting the Bottom and Top Acceleration Structure rebuild and AO.")
```
Bottom and Top **level** Acceleration Structure

gpx1000 committed:
fix typo in README.md

gpx1000 committed:
fix typo in README.md and in CMakeLists.txt

gpx1000 committed:
Merge branch 'master' into master

JoseEmilio-ARM commented on samples/extensions/raytracing_extended/raytracing_extended.cpp:
```diff
@@ -0,0 +1,1450 @@
+template <typename T>
+struct CopyBuffer
+{
+	std::vector<T> operator()(std::unordered_map<std::string, vkb::core::Buffer> &buffers, const char *bufferName)
```
style: snake_case for `bufferName` and `alreadyMapped`. Since this is also used by other samples (so far Ray Queries) maybe it could be moved to the framework, e.g. `common/utils.h`? Or would it be possible for them to use the `Buffer` class in `core/buffer.h`?

JoseEmilio-ARM commented on samples/extensions/raytracing_extended/raytracing_extended.cpp:
```diff
@@ -0,0 +1,1450 @@
+	VK_CHECK(vkCreateRayTracingPipelinesKHR(get_device().get_handle(), VK_NULL_HANDLE, VK_NULL_HANDLE, 1, &raytracing_pipeline_create_info, nullptr, &pipeline));
+}
+
+void RaytracingExtended::create_display_pipeline()
```
unused left-over?

JoseEmilio-ARM commented on samples/extensions/raytracing_extended/raytracing_extended.cpp:
```diff
@@ -0,0 +1,1450 @@
+
+void RaytracingExtended::draw_gui()
+{
+	gui->show_options_window(
```
Currently when running the sample I cannot see the GUI, is that expected?

gpx1000 replied:
until GUI's work in the framework, we took out the GUI here to resolve this.  However, will add it back when #334 is merged in.  Can do a separate PR for that addition.

gpx1000 replied:
GUI is removed until #334 is merged.

JoseEmilio-ARM commented on shaders/khr_ray_tracing_extended/raygen.rgen:
```diff
@@ -0,0 +1,170 @@
+	float tmax = 10000.0;
+
+	uint max_rays = maxRays;
+	max_rays = 100;
```
intended?

JoseEmilio-ARM commented on shaders/khr_ray_tracing_extended/raygen.rgen:
```diff
@@ -0,0 +1,170 @@
+  vec4 normal; // {nx, ny, nz, distance}
+};
+
+vec3 heatmap(float value, float minValue, float maxValue)
```
not used in this shader

JoseEmilio-ARM commented on shaders/khr_ray_tracing_extended/raygen.rgen:
```diff
@@ -0,0 +1,170 @@
+		object_type = uint(hitValue.intersection.w);
+		const vec3 object_intersection_pt = hitValue.intersection.xyz;
+		const vec3 object_normal = hitValue.normal.xyz;
+		if (render_mode != 0)
```
would it possible to use defines that match enum values for `render_mode` and `object_type`, to improve readability? or maybe comments would be enough?

mklefrancois commented:
I tried it with a NVIDIA GeForce RTX 3060Ti and it works.
* Should there be a user interface to change the drawing mode? It does not show up in my test.
* Is it possible that there is a problem with the animation update? The buffers are updated with map/unmap, but I don't see how that protects against updating the buffer while drawing, unless I'm missing a wait/fence somewhere.
* Should the animation be created with `VK_BUILD_ACCELERATION_STRUCTURE_ALLOW_UPDATE_BIT_KHR` to update it?
  * The BLAS update could then set both `srcAccelerationStructure` and `dstAccelerationStructure` to the same acceleration structure when it is recreated.
* Wouldn't it be a good idea to have the animation executed with a compute shader?

wflohry committed:
Updates for PR comments

wflohry committed:
Adding memory barriers

wflohry committed:
Updating shaders to use defines

gpx1000 committed:
Merge remote-tracking branch 'downstream/feature/raytrace-extended'

# Conflicts:
#	samples/extensions/raytracing_extended/raytracing_extended.cpp
#	shaders/khr_ray_tracing_extended/closesthit.rchit

gpx1000 committed:
clang format cleanup

gpx1000 committed:
add missing copyright to new header

gpx1000 commented:
> I tried it with a NVIDIA GeForce RTX 3060Ti and it works.
> 
> * Should there be a user interface to change the drawing mode? It does not show up in my test.
> * Is it possible that there is a problem with the animation update? The buffers are updated with map/unmap, but I don't see how that protects against updating the buffer while drawing, unless I'm missing a wait/fence somewhere.
> * Should the animation be created with `VK_BUILD_ACCELERATION_STRUCTURE_ALLOW_UPDATE_BIT_KHR` to update it?
>   
>   * The BLAS update could then set both `srcAccelerationStructure` and `dstAccelerationStructure` to the same acceleration structure when it is recreated.
> * Wouldn't it be a good idea to have the animation executed with a compute shader?

Thanks for reviewing @mklefrancois. 
* Until the framework supports GUI use in Ray Tracing pipelines #334 addresses that shortcoming.  We've removed the GUI all together.
* fixed
* That's not a bad idea.  Using the UPDATE instead of the BUILD to update the AS instead of rebuild it whole sale.  As I'm using this as the first ray tracing sample to use the framework's acceleration structure I'm going to opt for doing this in a future PR.  The framework will take a bit more thought to handle updates vs rebuilds. *** update: I just went ahead and fixed it.
* While I agree it's a good optimization to do all animation in a compute shader, I worry that it would add complexity to the sample in an area that doesn't highlight the general case of how to do animation with ray tracing pipelines.  Thus, while it's an excellent idea to do animation in a compute shader, it should probably be beyond the scope.

gpx1000 committed:
fix attribution in the tutorial.md

gpx1000 committed:
framework:
acceleration structure: (build) fix a bug which only would allow a single triangle to be included in a acceleration structure.
acceleration structure: (build) allow for more than one call to build (useful in animation).  This was caused by recreating the buffer each time rather than reusing it if it already exists.
acceleration structure: (add triangles) allow for developer to specify the data address of the vertex index and transform buffers that are used.  This allows for multi model gltf files to work as in the sponza scene.
acceleration structure: enable buffer retrieval to allow for developer to add in memory barriers.
acceleration structure: provide developer means to wipe the geometries added to the acceleration structure.  This is very important if animating and thus recalling the same structure(s).  NB: a future feature would be to index the geometries and instances to allow for updating existing data individually rather than a requirement to wipe out everything and rebuild everything.
scratch buffer: add a return for the scratch buffer's size to allow for reusing the scratch buffer if it's the same size as previously allocated.

Sample:
As this is the first ray tracing sample to make use of @Sascha William's excellent acceleration structure framework, and we already have the existing long form, I opted to enable a macro to turn on/off using the framework's acceleration structure.  I think this is valuable to end users to show a precise "what the frameork does" behind the scenes.  Future samples should opt to use the framework.

gpx1000 committed:
run clang-format over scratch_buffer.h

gpx1000 committed:
Add in proper update logic for the acceleration_structure in both the framework and the sample.

gpx1000 committed:
Ensure that only updated geometry are used in each update pass and make framework the default AS.

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Small changes for buffer.

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
Merge branch 'master-upstream'

JoseEmilio-ARM commented on samples/extensions/raytracing_extended/raytracing_extended.h:
```diff
@@ -21,7 +21,7 @@
 #pragma once
 
-#define USE_FRAMEWORK_ACCELERATION_STRUCTURE
+//#define USE_FRAMEWORK_ACCELERATION_STRUCTURE
```
why make it optional?

gpx1000 replied:
It's optional because it's the first sample to use the framework for acceleration structure.  I felt it's good in the first sample to show an equivalent how to build with and without the framework.  This can be both instructive and a nice validation of expectations.

gpx1000 committed:
Merge branch 'master-upstream'

mklefrancois reviewed (dismissed):
Tested once again and works fine

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
run clang-format

gpx1000 committed:
fix android

gpx1000 committed:
fix android

mklefrancois reviewed (dismissed):
Tested and works

marty-johnson59 commented:
@aurolou, can we get your help to review this RT sample as well as PR #317?  These are ray tracing samples and we'd really like to make sure they support AMD before releasing them.  Thanks!

aurolou commented:
This sample throws me an OUT_OF_POOL_MEMORY error when allocating descriptor sets in raytracing_extended.cpp, line 884. The Pool Size struct is missing some entries. If fixed, the sample works fine, tested on a RX 6800 XT.

JoseEmilio-ARM commented on framework/core/acceleration_structure.h:
```diff
@@ -56,16 +56,35 @@ class AccelerationStructure
+	                              uint64_t                            transform_buffer_data_address = 0);
 
 	/**
 	 * @brief Adds instance geometry to the acceleration structure (only valid for top level)
```
missing return documentation

gpx1000 committed:
Merge branch 'master-upstream'

gpx1000 committed:
address the returns doc.

gpx1000 committed:
Merge branch 'KhronosGroup:master' into master
